llvm/lib/IR/RepoDefinition.cpp

```diff
@@ -79,187 +88,241 @@ const Constant *getAliasee(const GlobalAlias *GA) {
-    if (GV.isDeclaration())
-      continue;
-    calculateGOInfo(&GV, GOIMap);
-    ++GONum.VarNum;
-  }
-  for (auto &Fn : M.functions()) {
-    if (Fn.isDeclaration())
-      continue;
-    calculateGOInfo(&Fn, GOIMap);
-    ++GONum.FuncNum;
-  }
-  return std::make_pair(std::move(GOIMap), std::move(GONum));
+DigestType ModuleHashGenerator::calculateDigest(const GlobalObject *GO,
+                                                bool HasRepoDefinitionMD) {
+  auto const Result = std::get<DigestType>(
+      updateDigestUseDependenciesAndContributions(GO, HasRepoDefinitionMD));
+  assert(Visited.empty() && "Visited is expected to be empty!");
+  return Result;
 }
 
-std::tuple<bool, unsigned, unsigned> generateRepoDefinitions(Module &M) {
-  bool Changed = false;
-  GONumber GONum;
-  GOStateMap Visited;
-  Visited.reserve(M.size() + M.getGlobalList().size());
-  GOInfoMap GOIMap;
+void ModuleHashGenerator::digestModule(Module &M) {
+  auto GOSize = M.size() + M.getGlobalList().size();
+  Visited.reserve(GOSize);
+  GOHashCache.reserve(GOSize);
+  Changed = false;
 
-  // Calculate the GOInfoMap and GONumber.
-  std::tie(GOIMap, GONum) = calculateGONumAndGOIMap(M);
-
-#ifndef NDEBUG
-  for (auto &GI : GOIMap) {
-    LLVM_DEBUG(dbgs() << "\nGO Name:" << GI.first->getName() << "\n");
-    LLVM_DEBUG(GI.second.dump());
-  }
-#endif
+  // Calculate the GOInfoMap.
+  calculateGONumAndGOIMap(M);
 
   // Update the GO's digest using the dependencies and the contributions.
   for (auto &GO : M.global_objects()) {
     if (GO.isDeclaration())
       continue;
-    Visited.clear();
-    MD5 Hash = MD5();
-    auto Helper = [](const GlobalObject *GO, MD5 &GOHash, GOInfoMap &GOIMap) {
-      return accumulateGOInitialDigestAndGetGODigestState(GO, GOHash, GOIMap);
-    };
-    updateDigestUseDependenciesAndContributions(&GO, Hash, Visited, GOIMap,
-                                                Helper);
-    MD5::MD5Result Digest;
-    Hash.final(Digest);
-    set(&GO, Digest);
-    Changed = true;
+    set(&GO, calculateDigest(&GO, false));
   }
+}
+
+DigestType
+ModuleHashGenerator::calculateDigestUseRepoDefinition(const GlobalObject *GO) {
+  auto const Result = std::get<DigestType>(
+      updateDigestUseDependenciesAndContributions(GO, true));
+  assert(Visited.empty() && "Visited is expected to be empty!");
+  return Result;
+}
 
-  return std::make_tuple(Changed, GONum.VarNum, GONum.FuncNum);
+bool generateRepoDefinitions(Module &M) {
+  ModuleHashGenerator HC;
+  HC.digestModule(M);
+  return HC.isChanged();
 }
 
 DigestType calculateDigest(const GlobalObject *GO) {
-  GOStateMap Visited;
-  MD5 Hash = MD5();
-  GOInfoMap GOIMap;
-
-  updateDigestUseDependenciesAndContributions(
-      GO, Hash, Visited, GOIMap,
-      accumulateGOFinalDigestOrInitialDigestAndGetGODigestState);
-  MD5::MD5Result Digest;
-  Hash.final(Digest);
-  return Digest;
+  ModuleHashGenerator HC;
+  return HC.calculateDigestUseRepoDefinition(GO);
```

Comment:
This could call   `return HC.calculateDigest(GO, true);`, then you could remove calculateDigestUseRepoDefinition()

Reply:
...unless there is a reason why ModuleHashGenerator::calculateDigest() shouldn't be made public.

Reply:
Good point, changed.